python/sglang/bench_offline_throughput.py

```diff
@@ -0,0 +1,160 @@
+        )
+        parser.add_argument(
+            "--result-filename", type=str, default=BenchArgs.result_filename
+        )
+        # graphing
+        parser.add_argument("--graph-sql", type=str, default=BenchArgs.graph_sql)
+        parser.add_argument(
+            "--graph-filename", type=str, default=BenchArgs.graph_filename
+        )
+
+    @classmethod
+    def from_cli_args(cls, args: argparse.Namespace):
+        # use the default value's type to case the args into correct types.
+        attrs = [(attr.name, type(attr.default)) for attr in dataclasses.fields(cls)]
+        return cls(
+            **{attr: attr_type(getattr(args, attr)) for attr, attr_type in attrs}
+        )
+
+
+def prepare_synthetic_inputs_for_throughput_test(
+    batch_size: int, input_len: int, output_len: int
+):
+    input_ids = [[1] * input_len for _ in range(batch_size)]
+    sampling_params = {
+        "temperature": 0,
+        "min_new_tokens": output_len,
+        "max_new_tokens": output_len,
+    }
+    return input_ids, sampling_params
+
+
+def throughput_test_once(
+    run_name: str,
+    engine: Engine,
+    reqs: Tuple[List[List[int]], Dict],
+    output_len: int,
+):
+    measurement_results = {
+        "run_name": run_name,
+        "batch_size": len(reqs[0]),
+        "input_len": len(reqs[0][0]),
+        "output_len": output_len,
+    }
+
+    st = time.perf_counter()
+    gen_out = engine.generate(input_ids=reqs[0], sampling_params=reqs[1])
+    latency = time.perf_counter() - st
+
+    measurement_results["total_latency"] = latency
+    measurement_results["throughput"] = (
+        (measurement_results["input_len"] + output_len)
+        * measurement_results["batch_size"]
+    ) / latency
+
+    print(
+        f"Throughput: BSZ {measurement_results['batch_size']} tokens, "
+        f"Num sequences {len(reqs[0])}, throughput: "
+        f"{measurement_results['throughput']} tokens/s"
+    )
+    return measurement_results
+
+
+def throughput_test(
+    server_args: ServerArgs,
+    bench_args: BenchArgs,
+):
+    engine = getEngine(**dataclasses.asdict(server_args))
+    if not engine:
+        raise ValueError("Please provide valid engine arguments")
+
+    warmup_reqs = prepare_synthetic_inputs_for_throughput_test(
+        bench_args.batch_size[0], bench_args.input_len[0], bench_args.output_len[0]
+    )
+
+    # Warm up
+    throughput_test_once("warmup", engine, warmup_reqs, bench_args.output_len[0])
+
+    result_list = []
+    for bs, il, ol in itertools.product(
+        bench_args.batch_size, bench_args.input_len, bench_args.output_len
+    ):
+        reqs = prepare_synthetic_inputs_for_throughput_test(bs, il, ol)
+        ret = throughput_test_once(
+            bench_args.run_name, engine, reqs, bench_args.output_len[0]
+        )
+        if ret is not None:
+            result_list.append(ret)
+
+    if bench_args.result_filename:
+        with jsonlines.open(bench_args.result_filename, "a") as f:
```

Comment:
use `fout.write(json.dumps(value) + "\n")` to remove the additional dependency of `jsonlines`